### ISSUE TITLE
Add support for any S3 compatible object storage

### ISSUE THREAD

**Issue by vancoykendall** [mosaicml/streaming#260]
I'm not using AWS for my s3 object storage so it isn't supported by StreamingDataset. However, the only change I need to make is hard coding the endpoint_url to my custom one in [streaming/base/storage/download.py:47](https://github.com/mosaicml/streaming/blob/6e6fd76f7c4e5f9c21233181a923bcb98de0e051/streaming/base/storage/download.py#LL47C49-L47C49)

```python
# original
s3 = session.client('s3', config=config)
# hard coded endpoint_url
s3 = session.client('s3', config=config, endpoint_url="<custom_endpoint>")
```

It seems like #255 and #68 address similar issues for gcs and r2, both of which are s3 compatible object stores using boto3. To me it feels like it would makes sense to just have a single download function for any s3 compatible object store that uses boto3 and give the user the ability to specify client parameters like the endpoint_url?

**Comment by karan6181** [comment]
Hi @vancoykendall, thanks for filing an issue. Your use-case is totally valid. To support this, it would require some interface change for upload and download functions but nothing too crazy that can't be done. Wondering, do you have a list of s3 (boto3) compatible storage ?

**Comment by vancoykendall** [comment]
@karan6181 Thanks for the quick response. I don't have a list of s3 compatible storage options, my company just setup our own object store in our own data centers using the s3 api. I don't know how common this is, so maybe just adding the ability to specify the endpoint_url for the StreamingDataset to use it enough. It seemed unnecessary to add a new download function and uploader for my use case, but it may make sense in general for others.

**Comment by karan6181** [comment]
@vancoykendall Yes, that make sense. We will take a look at this soon. Not sure how much bandwidth you have but If you are open to contribute then I can help you out by sharing some pointers. Please let us know. Thank you!

**Comment by karan6181** [comment]
@vancoykendall I have one question. what is your dataset UrI prefix looks like ? is it still `s3://` or something else?

**Comment by vancoykendall** [comment]
> what is your dataset UrI prefix looks like ? is it still s3:// or something else?

Yes my url prefix still uses s3://

> Not sure how much bandwidth you have but If you are open to contribute then I can help you out by sharing some pointers.

I don't have a ton of bandwidth, but would definitely be open to contribute and would appreciate some pointers.

**Comment by karan6181** [comment]
Hi @vancoykendall , the above feature has been merged as part of PR #265. Please let us know if you see any issues.
